includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -193,54 +201,15 @@ private static function determine_which_images_to_fetch( &$dimensions, &$urls_to
 	/**
 	 * Fetch dimensions of remote images
 	 *
-	 * @param array  $urls_to_fetch Image src urls to fetch.
-	 * @param array  $images Array to populate with results of image/dimension inspection.
-	 * @param string $mode Whether image dimensions should be extracted concurrently or synchronously.
-	 */
-	private static function fetch_images( $urls_to_fetch, &$images, $mode ) {
-		if ( 'synchronous' === $mode ||
-			false === function_exists( 'curl_multi_exec' ) ||
-			version_compare( PHP_VERSION, '5.4.0' ) < 0
-		) {
-			self::fetch_images_via_fast_image( $urls_to_fetch, $images );
-		} else {
-			self::fetch_images_via_faster_image( $urls_to_fetch, $images );
-		}
-	}
-
-	/**
-	 * Fetch images via FastImage library
-	 *
-	 * @param array $urls_to_fetch Image src urls to fetch.
-	 * @param array $images Array to populate with results of image/dimension inspection.
-	 */
-	private static function fetch_images_via_fast_image( $urls_to_fetch, &$images ) {
-
-		$image = new FastImage();
-		$urls  = array_keys( $urls_to_fetch );
-
-		foreach ( $urls as $url ) {
-			$result = $image->load( $url );
-			if ( false === $result ) {
-				$images[ $url ]['size'] = self::STATUS_IMAGE_EXTRACTION_FAILED;
-			} else {
-				$size = $image->getSize();
-
-				$images[ $url ]['size'] = $size;
-			}
-		}
-	}
-
-	/**
-	 * Fetch images via FasterImage library
+	 * @throws Exception When cURL handle cannot be added.
 	 *
 	 * @param array $urls_to_fetch Image src urls to fetch.
 	 * @param array $images Array to populate with results of image/dimension inspection.
 	 */
-	private static function fetch_images_via_faster_image( $urls_to_fetch, &$images ) {
+	private static function fetch_images( $urls_to_fetch, &$images ) {
 		$urls       = array_keys( $urls_to_fetch );
 		$user_agent = apply_filters( 'amp_extract_image_dimensions_get_user_agent', self::get_default_user_agent() );
-		$client     = new \FasterImage\FasterImage( $user_agent );
+		$client     = new \FasterImage\FasterImage( $user_agent ); // @todo The $user_agent is not actually able to be passed in this way to FasterImage. Needs another patch?
```

Comment:
This is something we should fix as well.

Reply:
@felixarntz What is the workflow to add a new patch?

I'm guessing we will want to do something like this here:

```diff
--- a/wp-content/plugins/amp/third_party/fasterimage/FasterImage.php
+++ b/wp-content/plugins/amp/third_party/fasterimage/FasterImage.php
@@ -25,7 +25,25 @@ class FasterImage
      */
     protected $timeout = 10;
 
-    /**
+	/**
+	 * User agent.
+	 *
+	 * @var string
+	 */
+	protected $user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.110 Safari/537.36';
+
+	/**
+	 * FasterImage constructor.
+	 *
+	 * @param string $user_agent User agent.
+	 */
+	public function __construct( $user_agent = '' ) {
+		if ( $user_agent ) {
+			$this->user_agent = $user_agent;
+		}
+	}
+
+	/**
      * Get the size of each of the urls in a list
      *
      * @param array $urls
@@ -115,7 +133,7 @@ class FasterImage
         curl_setopt($ch, CURLOPT_TIMEOUT, $this->timeout);
 
         #  Some web servers require the useragent to be not a bot. So we are liars.
-        curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/27.0.1453.110 Safari/537.36');
+        curl_setopt($ch, CURLOPT_USERAGENT, $this->user_agent);
         curl_setopt($ch, CURLOPT_HTTPHEADER, [
             "Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5",
             "Cache-Control: max-age=0",
```

Reply:
I fixed this now.

What I currently do is I have a local checkout of the FasterImage library in the tagged version that we're using. I apply the existing patch against it, change code, recreate the patch again.

Note that you can't just use the `master` branch since it contains code that isn't part of a release yet. However, for future releases, our patch file will just apply until something in that same area changes, which would naturally cause a merge conflict. Composer would notify us about that.